### PR TITLE
feat: add API to retrieve grant notes for specific user via Finder API

### LINKED ISSUE #3426: Follow+Notes: Retrieve note(s) for grant by a specific user via Finder API
Subtask of **[[STORY]: Update 'Status' to 'Follow + Note' feature](https://github.com/usdigitalresponse/usdr-gost/issues/2960)** #2960 

#### Blocked by

N/A

#### Blocks

- #3428

### Definition of Done

A new API route exists that allows retrieval of the latest revision for each note pertaining to a specific user and grant.

### Implementation Details

- In `packages/server/src/lib/grantsCollaboration/notes.js`, update the exported interface with the following functions:
  - `getOrganizationNotesForGrantByUser(knex, organizationId, userId, grantId, { afterRevision, limit = 50 } = {})`
    - Queries for the most-recent revision of each grant note pertaining to the identified grant and user.
    - Results should be ordered by `grant_notes_revisions.created_at` in descending order (most recent first).
    - To implement this functionality, I recommend the following approach (which avoids reimplementing database query behavior and maintains compatibility with existing functionality and associated tests):
      - Rename the existing `getOrganizationNotesForGrant()` function to `getCurrentNoteRevisions()` and update it's argument signature to `(knex, { grantId, organizationId, userId } = {}, { afterRevision, limit = 50 } = {})`.
      - Modify the body of the newly-renamed `getCurrentNotesRevisions()` function so that the `WHERE` filters of the query pertaining to `grant_notes.grant_id = :grantId` and `grant_notes.user_id = :userId` are conditional and only applied when `grantId` and `userId` are not null/`undefined`. In other words, we want to make the behavior of limiting query results to those pertaining to a specific user and/or grant _optional_.
      - Write the new `getOrganizationNotesForGrantByUser()` function as a simple wrapper that calls `getCurrentNotesRevisions()` and returns the return value from that internal call, e.g.
        ```js
        function getOrganizationNotesForGrantByUser(knex, grantId, organizationId, userId, { afterRevision, limit = 50 } = {}) {
            return getCurrentNoteRevisions(knex, { grantId, userId }, { afterRevision, limit });
        }
        ```
        This function should be exported by the `packages/server/src/lib/grantsCollaboration/notes.js` module.
      - Similarly, write a new exported `getOrganizationNotesForGrant(knex, grantId, organizationId, { afterRevision, limit = 50 } = {})` function (which implements the same original call signature) as another simple wrapper for `getCurrentNoteRevisions()`, e.g.
        ```js
        function getOrganizationNotesForGrant(knex, grantId, organizationId, { afterRevision, limit = 50 } = {}) {
            return getCurrentNoteRevisions(knex, { grantId, organizationId }, { afterRevision, limit });
        }
        ```
- In `packages/server/src/lib/grantsCollaboration/index.js`, import/export the `getOrganizationNotesForGrantByUser()` function exported from `packages/server/src/lib/grantsCollaboration/notes.js`
- In `packages/server/src/routes/grants.js`, define a new `GET /:grantId/notes/user/:userId` API route handler as follows:
  - Requires authentication.
  - Calls `getOrganizationNotesForGrantByUser()` from `packages/server/src/lib/grantsCollaboration` to retrieve a (paginated) result set of notes pertaining to the grant, organization, and user identified in the request.
    - The `grantId` and `userId` call arguments to `getOrganizationNotesForGrantByUser()` should all be derived from the corresponding request path parameter (`req.params`) values.
    - The `organizationId` call argument to `getOrganizationNotesForGrantByUser()` should be the `tenant_id` of the user provided by the request session (`req.session.user.tenant_id`)
    - If a `?paginateFrom` query parameter value is provided in the request, that value should be provided as the `afterRevision` argument in the call to `getOrganizationNotesForGrantByUser()`.
    - If a `?limit` query parameter value is provided in the request, that value (as a parsed integer)  should be provided as the `limit` argument in the call to `getOrganizationNotesForGrantByUser()`. Additionally, this value must pass the following validation checks (or else respond with a `400 Bad Request` status):
      - Can be successfully parsed as an integer
      - Parsed integer value must be greater than or equal to 1
      - Parsed integer value must not exceed 100
  - Following a successful call to `getOrganizationNotesForGrantByUser()`, provides the function's return value as a JSON-serialized response.

### PR DESCRIPTION
**Pull request by sushilrajeeva**
## Description
This PR implements the ability to retrieve notes for a specific user and grant via the Finder API, addressing the requirements outlined in the issue #3426 

Changes include :
  - Implemented `getOrganizationNotesForGrantByUser()` function in `grantsCollaboration/notes.js`
  - Renamed `getOrganizationNotesForGrant()` to `getCurrentNoteRevisions()` and updated signature
  - Made conditional `WHERE` clause for `grantId` and `userId` in `getCurrentNoteRevisions()`
  - Created API route `GET /:grantId/notes/user/:userId` in `grants.js` for fetching user-specific grant notes
  - Added validation for `afterRevision` and `limit` query parameters in the new API
  - Ensured results are paginated and ordered by `created_at` in descending order
  - Added necessary imports/exports in `grantsCollaboration/index.js`

Fixes #3426 

## Screenshots / Demo Video
N/A
### Automated and Unit Tests
- [ ] Added Unit tests

### Manual tests for Reviewer
- [ ] Added steps to test feature/functionality manually

## Checklist
- [X] Provided ticket and description
- [ ] Provided screenshots/demo
- [ ] Provided testing information
- [x] Provided adequate test coverage for all new code
- [X] Added PR reviewers

### CODE REVIEW

**Comment by greg-adams** [comment]
@sushilrajeeva is it possible to add test coverage for this new route?

**Review by TylerHendrickson — APPROVED** [review]
@sushilrajeeva Looks good!

**Review by TylerHendrickson — CHANGES_REQUESTED** [review]
@sushilrajeeva Apologies for the premature approval – I missed checking the unit tests. Agree with @greg-adams' [comment](https://github.com/usdigitalresponse/usdr-gost/pull/3471#issuecomment-2334940744) that it would be good for this new route to have test coverage.

File: packages/server/src/lib/grantsCollaboration/notes.js [hunk 1]
```diff
@@ -100,4 +117,28 @@ async function getOrganizationNotesForGrant(knex, grantId, organizationId, { aft
...
 
-module.exports = { saveNoteRevision, getOrganizationNotesForGrant };
+async function getOrganizationNotesForGrantByUser(
+    knex,
+    organizationId,
+    userId,
+    grantId,
+    { afterRevision, limit = 50 } = {},
+) {
+    return getCurrentNoteRevisions(knex, { grantId, userId }, { afterRevision, limit });
```
Comment (TylerHendrickson) [inline packages/server/src/lib/grantsCollaboration/notes.js:127]: Per https://github.com/usdigitalresponse/usdr-gost/issues/3426#issuecomment-2358790420:
```suggestion
    return getCurrentNoteRevisions(knex, { grantId, organizationId, userId }, { afterRevision, limit });
```

**Review by TylerHendrickson — APPROVED** [review]
As discussed in today's meeting, I think this is okay to merge given the lib tests are present. We will create a separate issue to address missing tests for `packages/server/routes/grants.js`.

**Pull request merged**